### PR TITLE
Add type fp32_fast_tf32 for gemm and corresponding deep relu example

### PR DESCRIPTION
**Pull request by amkatrutsa**

### CODE REVIEW

File: include/nntile/starpu/add_fiber.hh [hunk 1]
```diff
@@ -9,7 +9,8 @@
  *
  * @version 1.0.0
  * @author Aleksandr Mikhalev
- * @date 2023-07-21
+ * @author Aleksandr Mikhalev
+ * @date 2024-03-13
```
Comment (Muxas) [inline include/nntile/starpu/add_fiber.hh:13]: main branch has updated header: it no longer contains authors and dates. But we keep all the rest: copyrights, short NNTile description, file path with file description and version

File: include/nntile/starpu.hh [hunk 2]
```diff
@@ -113,7 +112,7 @@ void init()
     dgelutanh::init();
     drelu::init();
     gemm::init();
-    gemm_ex::init();
+    // gemm_ex::init();
```
Comment (Muxas) [inline include/nntile/starpu.hh:115]: you may remove this line and corresponding gemm_ex.hh file

File: include/nntile/starpu.hh [hunk 3]
```diff
@@ -184,7 +183,7 @@ void restrict_where(uint32_t where)
     dgelutanh::restrict_where(where);
     drelu::restrict_where(where);
     gemm::restrict_where(where);
-    gemm_ex::restrict_where(where);
+    // gemm_ex::restrict_where(where);
```
Comment (Muxas) [inline include/nntile/starpu.hh:186]: this is also not needed anymore

File: include/nntile/starpu.hh [hunk 4]
```diff
@@ -255,7 +254,7 @@ void restore_where()
     dgelutanh::restore_where();
     drelu::restore_where();
     gemm::restore_where();
-    gemm_ex::restore_where();
+    // gemm_ex::restore_where();
```
Comment (Muxas) [inline include/nntile/starpu.hh:257]: not needed

File: include/nntile/starpu/add_fiber.hh [hunk 5]
```diff
@@ -22,23 +22,23 @@ namespace nntile::starpu::add_fiber
 {
 
 //! Structure for arguments
-template<typename T>
+// template<typename T>
```
Comment (Muxas) [inline include/nntile/starpu/add_fiber.hh:25]: remove

File: include/nntile/starpu/add_slice.hh [hunk 6]
```diff
@@ -22,14 +22,14 @@ namespace nntile::starpu::add_slice
 {
 
 //! Structure for arguments
-template<typename T>
+// template<typename T>
```
Comment (Muxas) [inline include/nntile/starpu/add_slice.hh:25]: remove

File: include/nntile/starpu/subtract_indexed_outputs.hh [hunk 7]
```diff
@@ -25,7 +25,7 @@ struct args_t
 {
     Index n_labels;
     Index n_outputs;
-    T value;
+    scal_t value;
```
Comment (Muxas) [inline include/nntile/starpu/subtract_indexed_outputs.hh:28]: typename T above shall be removed

File: include/nntile/starpu/softmax.hh [hunk 8]
```diff
@@ -28,7 +28,7 @@ struct args_t
     Index m;
     Index n;
     Index k;
-    T alpha;
+    scal_t alpha;
```
Comment (Muxas) [inline include/nntile/starpu/softmax.hh:31]: template typename T above this line shall be removed

File: include/nntile/starpu/add.hh [hunk 9]
```diff
@@ -32,8 +32,8 @@ struct args_t
         {
         }
     Index nelems;
-    T alpha;
-    T beta;
+    scal_t alpha;
```
Comment (Muxas) [inline include/nntile/starpu/add.hh:35]: template line is to be removed

File: include/nntile/starpu/total_sum_accum.hh [hunk 10]
```diff
@@ -24,7 +24,7 @@ namespace nntile::starpu::total_sum_accum
 template<typename T>
 struct args_t
 {
-    T alpha;
+    scal_t alpha;
```
Comment (Muxas) [inline include/nntile/starpu/total_sum_accum.hh:27]: template<...> to be removed

File: src/starpu/accumulate.cc [hunk 11]
```diff
@@ -75,6 +75,23 @@ void init()
...
+            {cpu<fp32_t>},
+#ifdef NNTILE_USE_CUDA
+            {cuda<fp32_t>}
+#else // NNTILE_USE_CUDA
+            {}
+#endif // NNTILE_USE_CUDA
+            );
+    codelet_fp32_fast_tf32.nbuffers = 2;
+    codelet_fp32_fast_tf32.modes[0] = static_cast<starpu_data_access_mode>(
+            //STARPU_RW | STARPU_COMMUTE);
```
Comment (Muxas) [inline src/starpu/accumulate.cc:90]: Codelet allows A+BC+DE to be computed in (A+BC)+DE order and in (A+DE)+DC order. Uncomment line 90 and remove line 91.

File: src/starpu/accumulate_hypot.cc [hunk 12]
```diff
@@ -75,6 +75,22 @@ void init()
...
+            {cpu<fp32_t>},
+#ifdef NNTILE_USE_CUDA
+            {cuda<fp32_t>}
+#else // NNTILE_USE_CUDA
+            {}
+#endif // NNTILE_USE_CUDA
+            );
+    codelet_fp32_fast_tf32.nbuffers = 2;
+    codelet_fp32_fast_tf32.modes[0] = static_cast<starpu_data_access_mode>(
+            //STARPU_RW | STARPU_COMMUTE);
```
Comment (Muxas) [inline src/starpu/accumulate_hypot.cc:90]: uncomment, just like it is done above

File: src/starpu/accumulate_maxsumexp.cc [hunk 13]
```diff
@@ -75,6 +75,23 @@ void init()
...
+            {cpu<fp32_t>},
+#ifdef NNTILE_USE_CUDA
+            {cuda<fp32_t>}
+#else // NNTILE_USE_CUDA
+            {}
+#endif // NNTILE_USE_CUDA
+            );
+    codelet_fp32_fast_tf32.nbuffers = 2;
+    codelet_fp32_fast_tf32.modes[0] = static_cast<starpu_data_access_mode>(
+            //STARPU_RW | STARPU_COMMUTE);
```
Comment (Muxas) [inline src/starpu/accumulate_maxsumexp.cc:90]: uncomment

File: src/starpu/randn.cc [hunk 14]
```diff
@@ -96,6 +103,11 @@ void init()
             nullptr,
             {cpu_ndim0<fp32_t>},
             {});
+
+    codelet_fp32_fast_tf32_ndim0.init("nntile_randn_fp32_fats_tf32",
```
Comment (Muxas) [inline src/starpu/randn.cc:107]: fats->fast

Comment (Muxas) [reply src/starpu/randn.cc:107]: Try to `grep "fats" -R .` to find other possible typos

File: src/starpu/total_sum_accum.cc [hunk 15]
```diff
@@ -28,7 +28,7 @@ void cpu(void *buffers[], void *cl_args)
 #ifndef STARPU_SIMGRID // Run the code only if this is not a simulation
     // Get arguments
     auto args = reinterpret_cast<args_t<T> *>(cl_args);
-    T alpha = args->alpha;
+    scal_t alpha = args->alpha;
```
Comment (Muxas) [inline src/starpu/total_sum_accum.cc:31]: args shall not depend on template argument T (1 line above)

File: src/starpu/total_sum_accum.cc [hunk 16]
```diff
@@ -52,7 +52,7 @@ void cuda(void *buffers[], void *cl_args)
...
```
Comment (Muxas) [inline src/starpu/total_sum_accum.cc:55]: remove <T> argument away from args_t

File: src/starpu/total_sum_accum.cc [hunk 17]
```diff
@@ -98,6 +98,18 @@ void init()
             {}
 #endif // NNTILE_USE_CUDA
             );
+
+    codelet_fp32_fast_tf32.init("nntile_total_sum_accum_fp32_fast_tf32",
+            footprint<fp32_t>,
```
Comment (Muxas) [inline src/starpu/total_sum_accum.cc:103]: This footprint functions shall lose template argument

File: tests/starpu/gemm.cc.in
Comment (Muxas) [inline tests/starpu/gemm.cc.in:1]: Need to add a test with this new type

File: wrappers/python/nntile/__init__.py [hunk 18]
```diff
@@ -12,4 +12,4 @@
 # @version 1.0.0
 
 from .nntile_core import starpu, tile, TransOp, trans, notrans
-from . import layer, loss, model, tensor, pipeline, optimizer
+# from . import layer, loss, model, tensor, pipeline, optimizer
```
Comment (Muxas) [inline wrappers/python/nntile/__init__.py:15]: I mark this line to see if it is uncommented by the end of work to support TF32

File: wrappers/python/nntile/layer/__init__.py [hunk 19]
```diff
@@ -14,9 +14,8 @@
 from .base_layer import BaseLayer
 from .act import Act
 from .linear import Linear
-from .attention import Attention
-from .attention_single_head import AttentionSingleHead
```
Comment (Muxas) [inline wrappers/python/nntile/layer/__init__.py:18]: Add this line back when you finish with this work

File: wrappers/python/nntile/layer/__init__.py [hunk 20]
```diff
@@ -14,9 +14,8 @@
...
-from .flash_attention import FlashAttention
+# from .attention import Attention
```
Comment (Muxas) [inline wrappers/python/nntile/layer/__init__.py:17]: uncomment by end of work

File: wrappers/python/nntile/layer/__init__.py [hunk 21]
```diff
@@ -14,9 +14,8 @@
...
+# from .flash_attention import FlashAttention
```
Comment (Muxas) [inline wrappers/python/nntile/layer/__init__.py:18]: uncomment by end of work

File: wrappers/python/nntile/model/__init__.py [hunk 22]
```diff
@@ -15,5 +15,5 @@
 from .deep_linear import DeepLinear
 from .deep_relu import DeepReLU
 from .deep_relu_mp import DeepReLU_mp
-from .gpt2 import GPT2Config, GPT2Model
+# from .gpt2 import GPT2Config, GPT2Model
```
Comment (Muxas) [inline wrappers/python/nntile/model/__init__.py:18]: uncomment by end of work

File: wrappers/python/nntile/nntile_core.cc [hunk 23]
```diff
@@ -276,10 +277,71 @@ void tensor_from_array(const tensor::Tensor<T> &tensor,
     tensor.mpi_flush();
 }
 
+void tensor_from_array(const tensor::Tensor<fp32_fast_tf32_t> &tensor,
+        const py::array_t<fp32_t> &array)
+{
+    // Treat special 0-dimensional case, where NNTile assumes 1 element in a
```
Comment (Muxas) [inline wrappers/python/nntile/nntile_core.cc:283]: good for now, but maybe it is easier to make a mock conversion to fp32?
```
auto tensor_fp32 = reinterpret_cast<const tensor::Tensor<fp32_t> *>(&tensor);
tensor_from_array(*tensor_fp32, array);
```

File: wrappers/python/nntile/nntile_core.cc [hunk 24]
```diff
@@ -343,6 +405,66 @@ void tensor_to_array(const tensor::Tensor<T> &tensor,
     tmp.unregister();
 }
 
+void tensor_to_array(const tensor::Tensor<fp32_fast_tf32_t> &tensor, py::array_t<fp32_t> &array)
```
Comment (Muxas) [inline wrappers/python/nntile/nntile_core.cc:408]: the same with mocking data type can help to reduce code while maintaining readability

File: wrappers/python/nntile/nntile_core.cc [hunk 25]
```diff
@@ -408,36 +536,46 @@ void def_mod_tensor(py::module_ &m)
...
+    m.def("gemm_async_fp32_fast_tf32", &gemm_async<fp32_fast_tf32_t>);
+
+    m.def("gemm_fp64", &gemm<fp64_t>);
+    m.def("gemm_fp32", &gemm<fp32_t>);
+    m.def("gemm_fp16", &gemm<fp16_t>);
+    m.def("gemm_fp32_fast_tf32", &gemm<fp32_fast_tf32_t>);
     // Mixed precision gemm (FP32_FAST_FP16)
-    m.def("gemm_ex_async_fp32", &gemm_ex_async<fp32_t>);
-    m.def("gemm_ex_fp32", &gemm_ex<fp32_t>);
+    // m.def("gemm_ex_async_fp32", &gemm_ex_async<fp32_t>);
```
Comment (Muxas) [inline wrappers/python/nntile/nntile_core.cc:555]: remove these lines, they are no longer needed

File: wrappers/python/nntile/tensor.py [hunk 26]
```diff
@@ -62,27 +62,32 @@ def gemm_async(alpha: float, trans_A: TransOp, A: Tensor, trans_B: TransOp, \
...
-        trans_B: TransOp, B: Tensor, beta: float, C: Tensor, ndim: int, \
-        batch_ndim: int, redux: int=0) -> None:
-    if type(A) is not type(B) or type(A) is not type(C):
-        raise TypeError
-    if type(A) is core_tensor.Tensor_fp32:
-        core_tensor.gemm_ex_async_fp32(alpha, trans_A, A, trans_B, B, beta, C,
-                ndim, batch_ndim, redux)
-    else:
-        raise TypeError
+# def gemm_ex_async(alpha: float, trans_A: TransOp, A: Tensor, \
```
Comment (Muxas) [inline wrappers/python/nntile/tensor.py:72]: just remove it, as this operation is no more available

File: src/starpu/randn.cc [hunk 27]
```diff
@@ -81,13 +81,20 @@ uint32_t footprint(struct starpu_task *task)
 }
 
 Codelet codelet_fp32, codelet_fp64, codelet_fp32_ndim0, codelet_fp64_ndim0;
+Codelet codelet_fp32_fast_tf32, codelet_fp32_fast_tf32_ndim0;
```
Comment (Muxas) [inline src/starpu/randn.cc:84]: `mean` and `stddev` shall be of `scal_t` type everywhere. Just like `alpha` and `beta` parameters in other functions.

File: src/starpu/subtract_indexed_outputs.cc [hunk 28]
```diff
@@ -71,7 +71,7 @@ static
 uint32_t footprint(struct starpu_task *task)
 {
     // Get arguments
-    auto args = reinterpret_cast<args_t<T> *>(task->cl_arg);
+    auto args = reinterpret_cast<args_t *>(task->cl_arg);
```
Comment (Muxas) [inline src/starpu/subtract_indexed_outputs.cc:74]: this footprint functions shall not depend on <T>

Comment (Muxas) [reply src/starpu/subtract_indexed_outputs.cc:74]: Please check if all footprint functions are without template arguments now

File: wrappers/python/nntile/nntile_core.cc [hunk 29]
```diff
@@ -205,13 +205,14 @@ void def_mod_tile(py::module_ &m)
...
     def_class_tile<fp32_t>(m, "Tile_fp32");
+    def_class_tile<fp32_fast_tf32_t>(m, "Tile_fp32_fast_tf32");
     def_class_tile<fp64_t>(m, "Tile_fp64");
 }
 
 // numpy.ndarray -> Tensor
 template<typename T>
 void tensor_from_array(const tensor::Tensor<T> &tensor,
-        const py::array_t<T, py::array::f_style | py::array::forcecast> &array)
+        const py::array_t<T> &array)
```
Comment (Muxas) [inline wrappers/python/nntile/nntile_core.cc:215]: why did you remove `forcecast` and `f_style` flags? Without these flags copying data from numpy will be erroneous if user provides C-style input.

File: wrappers/python/nntile/nntile_core.cc [hunk 30]
```diff
@@ -276,10 +277,71 @@ void tensor_from_array(const tensor::Tensor<T> &tensor,
     tensor.mpi_flush();
 }
 
+void tensor_from_array(const tensor::Tensor<fp32_fast_tf32_t> &tensor,
+        const py::array_t<fp32_t> &array)
```
Comment (Muxas) [inline wrappers/python/nntile/nntile_core.cc:281]: The same with `forcecast` and `f_style`

File: wrappers/python/nntile/nntile_core.cc [hunk 31]
```diff
@@ -276,10 +277,71 @@ void tensor_from_array(const tensor::Tensor<T> &tensor,
...
+    tensor::scatter<fp32_fast_tf32_t>(tmp, tensor);
+    tmp.unregister();
+    tensor.mpi_flush();
+}
+
 // Tensor -> numpy.ndarray
 template<typename T>
-void tensor_to_array(const tensor::Tensor<T> &tensor,
-        py::array_t<T, py::array::f_style> &array)
+void tensor_to_array(const tensor::Tensor<T> &tensor, py::array_t<T> &array)
```
Comment (Muxas) [inline wrappers/python/nntile/nntile_core.cc:344]: Without `f_style` flag data will be written in a wrong way if the provided output is C-style.

File: wrappers/python/nntile/nntile_core.cc (hunk 24 above)
Comment (Muxas) [inline wrappers/python/nntile/nntile_core.cc:408]: `f_style` flag is missing

File: wrappers/python/nntile/nntile_core.cc [hunk 32]
```diff
@@ -357,8 +479,13 @@ void def_class_tensor(py::module_ &m, const char *name)
         def("invalidate_submit", &Tensor<T>::invalidate_submit).
         //def("invalidate_submit", &Tensor<T>::wont_use).
         def("wont_use", &Tensor<T>::wont_use).
-        def("from_array", tensor_from_array<T>).
-        def("to_array", tensor_to_array<T>).
+        // def("from_array", &tensor_from_array<T>).
+        def("from_array", [](const tensor::Tensor<fp32_t> & t, const py::array_t<fp32_t> & a) { return tensor_from_array<fp32_t>(t, a); } ).
```
Comment (Muxas) [inline wrappers/python/nntile/nntile_core.cc:483]: once flags are returned back these lines shall be updated

**Review by Muxas — COMMENTED** [review]
You shall proceed with the GPT2 now

File: wrappers/python/nntile/nntile_core.cc [hunk 33]
```diff
@@ -459,15 +460,19 @@ void def_mod_tensor(py::module_ &m)
     m.def("drelu_fp64", &drelu<fp64_t>);
     m.def("drelu_fp32", &drelu<fp32_t>);
     // Add other functions for Tensor<T>
-    m.def("fill_async_fp64", &fill_async<fp64_t>);
-    m.def("fill_async_fp32", &fill_async<fp32_t>);
-    m.def("fill_fp64", &fill<fp64_t>);
-    m.def("fill_fp32", &fill<fp32_t>);
+    m.def("fill_async_fp64", [](fp64_t val, const Tensor<fp64_t>& A) {return fill_async<fp64_t>(val, A); } );
+    m.def("fill_async_fp32", [](fp32_t val, const Tensor<fp32_t>& A) {return fill_async<fp32_t>(val, A); } );
+    m.def("fill_async_fp32_fast_tf32", [](fp32_t val, const Tensor<fp32_fast_tf32_t>& A) { auto t_fp32 = reinterpret_cast<const Tensor<fp32_t>* >(&A); return fill_async<fp32_t>(val, *t_fp32); });
```
Comment (amkatrutsa) [inline wrappers/python/nntile/nntile_core.cc:465]: @Muxas , please check if such adjusting of the fill function for the tf32 type is ok.

Comment (Muxas) [reply wrappers/python/nntile/nntile_core.cc:465]: Seems legit

File: wrappers/python/nntile/tensor.py [hunk 34]
```diff
@@ -551,9 +573,12 @@ def copy_intersection_async(x: TensorFloatOrInt, x_offset: List[int], \
 # Wrapper for multiprecision copy
 def copy_async(x: TensorFloatOrInt, y: TensorFloatOrInt) -> None:
     if type(x) is not type(y):
+        print("Type x = {}, type y = {}".format(type(x), type(y)))
```
Comment (Muxas) [inline wrappers/python/nntile/tensor.py:576]: Do not forget to remove this later

File: wrappers/python/nntile/model/gpt2.py [hunk 35]
```diff
@@ -12,10 +12,10 @@
 # @version 1.0.0
 
 from nntile.tensor import TensorTraits, Tensor, TensorOrNone, TensorMoments, \
-        notrans, trans, Tensor_fp32, Tensor_int64, Tensor_bool
+        notrans, trans, Tensor_fp32, Tensor_fp32_fast_tf32, Tensor_int64, Tensor_bool
 from nntile.model.base_model import BaseModel
-from nntile.layer import Linear, Embedding, AddSlice, LayerNorm, Attention, \
-        FlashAttention, Act, AttentionSingleHead
+from nntile.layer import Linear, Embedding, AddSlice, LayerNorm, Attention, Act
+# from nntile.layer import FlashAttention, AttentionSingleHead
```
Comment (Muxas) [inline wrappers/python/nntile/model/gpt2.py:18]: Do not forget to uncomment this later

File: src/starpu/axpy.cc.in [hunk 36]
```diff
@@ -139,7 +139,8 @@ void cuda2(void *buffers[], void *cl_args)
 {
 #ifndef STARPU_SIMGRID // Run the code only if this is not a simulation
     // Get arguments
-    auto args = reinterpret_cast<args2_t<T> *>(cl_args);
+    auto args = reinterpret_cast<args2_t *>(cl_args);
+    const T* alpha = reinterpret_cast<const T*>(&args->alpha);
```
Comment (amkatrutsa) [inline src/starpu/axpy.cc.in:143]: @Muxas  Is it OK to cast the pointer for scal_t to the proper type for compatibility with the original blas functions?

File: src/kernel/sum_slice/cuda.cu [hunk 37]
```diff
@@ -26,7 +26,7 @@ void cuda_kernel(Index m, Index n, Index k, Index mk, T alpha, const T *src,
     Index i0 = threadIdx.x + blockIdx.x*blockDim.x,
           i1 = threadIdx.y + blockIdx.y*blockDim.y;
     Index i2_start = threadIdx.z, i2_step = blockDim.z;
-    constexpr T zero = 0;
+    constexpr scal_t zero = 0;
```
Comment (Muxas) [inline src/kernel/sum_slice/cuda.cu:29]: This is incorrect, to be honest. But for now it is Ok. I mean in case of fp64_t or bf16_t this type shall not be scal_t.

File: src/kernel/sum_slice/cpu.cc [hunk 38]
```diff
@@ -38,7 +38,7 @@ void cpu(Index m, Index n, Index k, T alpha, const T *src, T beta, T *dst)
  * */
 {
     const Index mk = m * k;
-    constexpr T zero = 0;
+    constexpr scal_t zero = 0;
```
Comment (Muxas) [inline src/kernel/sum_slice/cpu.cc:41]: Revert back to type T

File: src/kernel/sum_slice/cuda.cu (hunk 37 above)
Comment (Muxas) [reply src/kernel/sum_slice/cuda.cu:29]: Actually, revert it back to type T

**Comment by Muxas** [comment]
This PR has moved to #60
